### ISSUE TITLE
Warning ``possibly-undefined-variable`` if a for loop variable that was not defined prior to the loop is accessed out of its intended scope

### ISSUE THREAD

**Issue by DetachHead** [PyCQA/pylint#5146]
### Current problem

an annoying feature in python is how variables are still in scope when you wouldn't expect them to be
```py
for value in [1,2,3]:
    print(value)

for other_value in [2,3,4]:
    print(value) # no warning
```

### Desired solution

a rule that warns when a variable is being accessed outside of the "scope" where it was defined, suggesting that you `del` the variable. in the example above, it could suggest the following fix:
```py
for value in [1,2,3]:
    print(value)
del value

for other_value in [2,3,4]:
    print(value) # error
```

**Comment by DudeNr33** [comment]
In essence this would mean you always have to put a ``del`` behind a for loop. 

Consider the following:
If you made the mistake to reference the loop variable again, ``pylint`` would complain (as soon as the new check is implemented, at least). 

What would be the next step? 
Adding ``del`` would not really help, the main problem is the fact that you accessed the loop variable. You have to change that (even more so because with the ``del`` your program would crash otherwise). 

Now that you changed it, the ``del`` no longer has any effect beside preventing this mistake from happening again. 

But if you are afraid of that, you might also be afraid that accessing loop variables after the loop finished could happen _anywhere_ in your code base.
So, should you ``del`` _all_ loop variables everytime and everywhere?
I don't think so. 

I'd also argue that this new checker should be optional. It can very well be intended to access the last value of a loop variable later on.

**Comment by Pierre-Sassoulas** [comment]
Here's [an article I found interesting](https://eli.thegreenplace.net/2015/the-scope-of-index-variables-in-pythons-for-loops/) which point out that there's also possible complications to using del (exceptions handling), provide historical context supposedly from Guido but without actual source or I would have linked that, and show some useful pattern that can be used because of this, like:

```python
for i in somegenerator():
    if isinteresing(i):
        break
dostuffwith(i)
````
Or 
````python
for i, item in enumerate(somegenerator()):
    dostuffwith(i, item)
print('The loop executed {0} times!'.format(i+1))
````

**Comment by DudeNr33** [comment]
In my first comment I initially had a sentence also saying that this would be perfectly valid code, but deleted it shortly after. 
The reason is that your iterable could be empty, and in that case the loop variable is undefined. 
IIRC PyCharm also flags this with something like "potentially undefined variable", and a warning from pylint would make sense. 
The solution would be to intialize the variable before the loop to make sure that it will be defined even if the iterable is empty.

**Comment by Pierre-Sassoulas** [comment]
I like your argument. 

``j`` is also defined for the second ``reused-loop-variable-out-of-scope`` (only it's before the first loop). Maybe it's the only valid case of ``# [reused-loop-variable-out-of-scope]``, because reusing it after another for loop seems weird ?

**Comment by Zeckie** [comment]
If you want to be explicit that you want the value available outside the loop, you could use something like:
```
k = 0
for j, item in enumerate(somegenerator(), start=1):
    k = j
    dostuffwith(j, item)
print(f'The first loop executed {k} times!')


for j, item in enumerate(somegenerator(), start=1):
    k = j
    dostuffwith(j, item)
print(f'One of the loops executed {k} times!')
```
But that doesn't solve the issue around whether the second loop expected `k` to be initialized again. However, the same could happen without for loops (especially in long functions).

I agree with DudeNr33 (and pycharm) that the warning should only occur when the variable has not been assigned before the loop, so could be undefined after.

**Comment by Pierre-Sassoulas** [comment]
Closing as pylint complain if the list is empty or can be empty.

```python
for value in [1, 2, 3]:
    print(value)

for other_value in [2, 3, 4]:
    print(value)  # no warning

for value_2 in []:
    print(value)

for other_value in [2, 3, 4]:
    print(value_2)  # [undefined-loop-variable]


def print_names(names):
    for name in names:
        print(name)
    send_message(name)  # [undefined-loop-variable]


def send_message(i):
    print(i)
```

Result:
```
************* Module a
a.py:11:10: W0631: Using possibly undefined loop variable 'value_2' (undefined-loop-variable)
a.py:17:17: W0631: Using possibly undefined loop variable 'name' (undefined-loop-variable)

------------------------------------------------------------------
Your code has been rated at 8.57/10 (previous run: 9.29/10, -0.71)
```
I think this is very close to the optimal in term of false positive/false negative. Thanks to @jacobtylerwalls for making this check very good in countless MR !
